src/erlcloud_kinesis.erl

```diff
@@ -130,6 +155,100 @@ delete_stream(StreamName, Config) when is_record(Config, aws_config) ->
+%%                 {<<"StartingHashKey">>, <<"34028236692093846346337460743176821145">>}
+%%             ]},
+%%             {<<"SequenceNumberRange">>, [
+%%                 {<<"StartingSequenceNumber">>, <<"49579844037727333356165064238440708846556371693205002258">>}
+%%             ]}
+%%         ], [
+%%             {<<"ShardId">>, <<"shardId-000000000002">>},
+%%             {<<"HashKeyRange">>, [
+%%                 {<<"EndingHashKey">>, <<"102084710076281539039012382229530463436">>},
+%%                 {<<"StartingHashKey">>, <<"68056473384187692692674921486353642281">>}
+%%             ]},
+%%             {<<"SequenceNumberRange">>, [
+%%                 {<<"StartingSequenceNumber">>, <<"49579844037749634101363594861582244564829020124710982690">>}
+%%             ]}
+%%         ], [
+%%             {<<"ShardId">>, <<"shardId-000000000003">>},
+%%             {<<"HashKeyRange">>, [
+%%                 {<<"EndingHashKey">>, <<"136112946768375385385349842972707284581">>},
+%%                 {<<"StartingHashKey">>, <<"102084710076281539039012382229530463437">>}
+%%             ]},
+%%             {<<"SequenceNumberRange">>, [
+%%                 {<<"StartingSequenceNumber">>, <<"49579844037771934846562125484723780283101668556216963122">>}
+%%             ]}
+%%         ]
+%%    ]}
+%% ]}
+%% '
+%%
+%% @end
+%%------------------------------------------------------------------------------
+
+-spec list_shards(binary()) -> erlcloud_kinesis_impl:json_return() | {error, any()}.
+list_shards(StreamName) ->
+   list_shards(StreamName, default_config()).
+
+-spec list_shards(binary(), list_shards_opts() | aws_config()) -> erlcloud_kinesis_impl:json_return().
+list_shards(StreamName, Config) when is_record(Config, aws_config) ->
+    list_shards(StreamName, [], Config);
+list_shards(StreamName, Options) ->
+    list_shards(StreamName, Options, default_config()).
+
+-spec list_shards(binary(), list_shards_opts(), aws_config()) -> erlcloud_kinesis_impl:json_return().
+list_shards(StreamName, Options, Config) when is_record(Config, aws_config) ->
+    DynamizedOptionsOrError = case lists:keymember(next_token, 1, Options) of
+        false ->
+            dynamize_options(Options);
+        true ->
+            case lists:keymember(stream_creation_timestamp, 1, Options) of
+                false ->
+                    dynamize_options(Options);
+                true ->
+                    {error, {incompatible_options, [next_token, stream_creation_timestamp]}}
+            end,
+            case lists:keymember(exclusive_start_shard_id, 1, Options) of
+                false ->
+                    dynamize_options(Options);
+                true ->
+                    {error, {incompatible_options, [next_token, exclusive_start_shard_id]}}
+            end
+    end,
+    case is_list(DynamizedOptionsOrError) of
+        true ->
+            Json = [{<<"StreamName">>, StreamName} | DynamizedOptionsOrError],
```

Comment:
Need to make this list further up (line 229), i.e. whatever `DynamizedOptionsOrError` is here if it's a list is good to go, no?

Reply:
1) does not address stream name + next token issue reported by @jlbarrett 
2) DynamizedOptionsOrError = dynamize_options(Options) can have `{error, {invalid...}}` and that is not caught here. 

I would suggest handling only two cases: 
- (next_token + max_result) and ignore other options. 
- any other combination and let user just get AWS error back.

Reply:
IMO you should discard all this error checking nonsense, AWS can check its own errors (as is true for all the other APIs here). i.e. just this:

```erlang
-spec list_shards(binary(), list_shards_opts(), aws_config()) -> erlcloud_kinesis_impl:json_return().
list_shards(StreamName, Options, Config) when is_record(Config, aws_config) ->
    StreamOpts = case lists:keymember(next_token, 1, Options) of
                     true -> [];
                     _ -> [{<<"StreamName">>, StreamName}]
                 end,
    Json = StreamOpts ++ dynamize_options(Options),
    erlcloud_kinesis_impl:request(Config, "Kinesis_20131202.ListShards", Json).
```